### ISSUE TITLE
define the terminal graphics protocol of our vile dreams

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#1734]
Having played around with terminal graphics protocols now about as much as any man living, it's time to propose one of our own, one designed with our goals in mind. I've started work on STEGAP here:

https://nick-black.com/dankwiki/index.php?title=Spriteful_TErminal_GrAphics_Protocol

It would be nice to perhaps implement it as well, in the hope of seeing it picked up.

Maybe the spec should live in Notcurses, or at least in git somewhere, to facilitate comments/changes.

**Comment by WSLUser** [comment]
When you have something cobbled enough together, there is a spot you know about to go to see about adoption and fleshing out particular implementation details: https://gitlab.freedesktop.org/terminal-wg/specifications. Kitty's graphic protocol is already radically updated from Sixel. So the question is do we really need another or could you just make some adjustments to Kitty's existing graphic protocol?

**Comment by dankamongmen** [comment]
@WSLUser kitty is definitely a close relation to my proposal. @kovidgoyal's brought things far forward with it. i do not think my proposal so far from the Kitty protocol that they couldn't be reconciled. STEGAP is certainly more kitty-like than it is sixel- or jexer-like. i hesitate to call iterm2 a protocol.

the biggest single inadequacy of Sixel from my perspective is ridding oneself of it. being able to reference graphics and delete/move them is 75% of the way there, and kitty has that for sure.

i think in most areas, STEGAP is locking down behaviors that kitty leaves unspecified, though this is not at all true everywhere.

one thing i just learned looking over Paul William's https://vt100.net/emu/dec_ansi_parser (as brought to my attention by @klamonte, on a thread in which she's talking mainly to @kovidgoyal, but from years ago lol, world keeps on spinnin') is that there are several byte values beyond 0x1b that one might need avoid throughout even bulk data, which makes me very sad. i'd love to get rid of the base64 business. it adds insult to injury.

**Comment by kovidgoyal** [comment]
Why not just specify these unspecified behavior's in kitty's protocol. I am happy to evolve it as needed, while keeping things backward compatible as far as possible. Though in case of severe need I am willing to break backwards compat.

**Comment by dankamongmen** [comment]
> Why not just specify these unspecified behavior's in kitty's protocol. I am happy to evolve it as needed, while keeping things backward compatible as far as possible. Though in case of severe need I am willing to break backwards compat.

i was getting all my thoughts down, and it ended up looking a good deal like your protocol. i wasn't sure it was going to =].

put another way: i set out to define my perfect protocol, and discovered it was not very far from what kitty's giving me. so yeah, i think running some suggestions by you will be sufficient, having put my thoughts together.

**Comment by kovidgoyal** [comment]
Three cheers for convergent evolution!

**Comment by dankamongmen** [comment]
i need to study the kitty animation extensions added in recent development. right now, all multiframe media gets rendered as successive entire frames, which is a tremendous bandwidth suck. i'm looking hard at encoding on the fly to PNG and transmitting that; i can burn the extra cycles on my side to convert, as they'll be more than paid back by reduced bandwidth. the encoding i propose in STEGAP just brings you back to baseline RGBA (from base64(RGBA)), but is still better than base64 imho if you have to send raw pixels.

the deflate component makes little sense to me for containerized images, as they presumably have domain-specific compression already. as i recall, it goes underneath the base64, so the 6 extra bits of pure redundancy can't be knocked out that way =[

but i think if i can use the animation protocol well, that'll actually be the big win, at least for the use case of e.g. video. need to take a look and grok it soon.

**Comment by kovidgoyal** [comment]
On Wed, Jun 09, 2021 at 08:56:31PM -0700, Nick Black wrote:
> i need to study the kitty animation extensions added in recent development. right now, all multiframe media gets rendered as successive entire frames, which is a tremendous bandwidth suck. i'm looking hard at encoding on the fly to PNG and transmitting that; i can burn the extra cycles on my side to convert, as they'll be more than paid back by reduced bandwidth. the encoding i propose in STEGAP just brings you back to baseline RGBA (from base64(RGBA)), but is still better than base64 imho if you have to send raw pixels.

I dont think one can robustly transmit arbitrary binary data, so some
form of encoding is required. I am happy to consider more efficient
forms as protocol extensions. But note that there will be a whole passel
of bytes it wont be safe to transmit. You will need to hunt this down
through both specs and actual testing with real world terminals. To me
that much effort is not worth it, so I just went with base64.

> 
> the deflate component makes little sense to me for containerized images, as they presumably have domain-specific compression already. as i recall, it goes underneath the base64, so the 6 extra bits of pure redundancy can't be knocked out that way =[

deflate is for use with RGB(A) data. 

> 
> but i think if i can use the animation protocol well, that'll actually be the big win, at least for the use case of e.g. video. need to take a look and grok it soon.

You should be able to, it allows you to build up a frame by transmitting
individual delta rectangles for the parts that have changed compared to
previous frames. It's essentially protocolized GIF, if you are familiar
with GIF.

**Comment by dankamongmen** [comment]
i'd also be very pleased were images to be scaled along with font size changes. a font size change, even if i'm notified with `SIGWINCH` when actually running, can screw up my output after i've stopped:

![2021-06-10-014125_1072x1417_scrot](https://user-images.githubusercontent.com/143473/121470992-f9e3ea80-c98c-11eb-8694-aed8b941e3da.png)

and there's nothing that can be done by my app at that point. if there's a good reason why you're not, i'd be interested in hearing it.

**Comment by kovidgoyal** [comment]
The main reason I have not addressed resize semantics (and this applies to all kinds of resizes, not just font size chages) is because it is far from clear what is the correct/desired semantics.  Lets consider some scenarios:

1) Screen is shrunk
  - should all image placements be shrunk accordingly?
  - how should their positioning be affected? should they remain anchored at the cell on which they started? For example, if an application intended an image to be horizontally centered, desired semantics would be to have the post resize image be centered as well, which it may no longer be.

2) Screen is expanded
 - should images be grown, even if that makes them pixelated/less clear?
 - same positioning issues

And coming to font size based resizes, resizing images in this scenario works if the intended semantics of the image are that it maintain a fixed ratio to the text size, but that's not always the desired behavior. Think for example of borders, lines, etc.

Basically, the space of desired behaviors on resize is very large. If we just pick one answer that will be wrong quite often. If instead we try to encode this space into the protocol so clients can tell the terminal how they want resizes handled for a given image, then that is a pretty large additional complexity to an already fairly complex protocol. One I didnt want to deal with.

That said if it is important to you then, I am happy to discuss further.

**Comment by dankamongmen** [comment]
ok, closing this bug as i agree that the kitty protocol is a sufficiently sound basis for work going forward, and i have no desire to add a new protocol for vanity's sake. further work will continue in the kitty issue tracker.
